### ISSUE TITLE
Add a new <syntheticIndexes/> flag to the code generator configuration

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#7858]
In jOOQ 3.4, we've added `<syntheticPrimaryKeys>` to the code generator configuration: #3154. This allowed for producing extra, "synthetic" meta data which would behave just the same way as actual primary key definitions.

It would be useful to also have this feature for "synthetic indexes"

**Comment by lukaseder** [comment]
If we want to support all index features, this turns out to be more complex than originally thought. We'd have to support:

- `WHERE` clauses
- `INCLUDE` clauses
- `ASC` / `DESC` clauses
- column expressions

**Comment by lukaseder** [comment]
Let's wait for a customer use case

**Comment by alf** [comment]
Happy to oblige.

Our strategy of mapping our GraphQL schema against code generated by jOOQ is working really well. We're now adding support for mapping a field in our GraphQL schema against our indexes so we can generate filtering and sorting.

Unfortunately, jOOQ does not generate anything for our indexes since they are in a different schema. This follows from our choice of using a schema with views against our actual schema. We'd now like to lift our indexes into the mix. Having support for `syntheticIndexes` would be very useful.

**Comment by lukaseder** [comment]
Thanks for the feedback, @alf.

One reason why this hasn't been implemented yet is because index meta data isn't really used inside of jOOQ itself, so the likelihood of this being very useful is small, as opposed to some other way of managing such meta data. Obviously, having all meta data in one place is neat, but can you confirm you can currently add another way of registering meta data on your end to implement your use-case?

**Comment by alf** [comment]
Yes, we can work around this by adding an extra configuration file that lets us extend what we can get from jOOQ.

We’ll build this feature now on top of the indexes jOOQ generates for us on our test database. Then we’ll see how we proceed.

I experience that the view layer inside the database isn’t as useful as first imagined, so we might also just open up the possibility of letting jOOQ access the underlying schema directly.

**Comment by alf** [comment]
I might have formulated myself poorly here.

What I mean is that I now consider it better to let jOOQ generate code from the underlying schemas directly, as opposed to having a schema of views in between. We will still use views when necessary, but working directly against the tables is simpler and thus preferable.

Much of our pain relating to views is caused by Oracle not supporting the RETURNING clause when using INSTEAD OF triggers. There are some other gotchas as well, but they are solvable.

Even if we didn’t have these problems, using the tables directly feels perfectly safe because jOOQ gives us great control of how our backend code uses our database. For example:

- We can communicate deprecation of tables and fields using comments in the database and set rules in our CI that break the build if these objects are being used.
- We can use Archunit and other tools in the Java ecosystem to set rules for who can use which database objects.
- jOOQ gives us a lot of possibilities for hooking into how the database is used, for instance with record listeners and other powerful interfaces for query rewriting etc

**Comment by lukaseder** [comment]
Thanks for your additional comments. I'm glad to see that you can work with the tables directly.

> * We can communicate deprecation of tables and fields using comments in the database and set rules in our CI that break the build if these objects are being used.

I reckon that this feature request will also be of interest, then:

- https://github.com/jOOQ/jOOQ/issues/9626

> * We can use Archunit and other tools in the Java ecosystem to set rules for who can use which database objects.

That sounds like an interesting blog post!
